core/shared/src/main/scala/scalaz/zio/Promise.scala

```diff
@@ -112,6 +112,26 @@ class Promise[E, A] private (private val state: AtomicReference[State[E, A]]) ex
       action
     })
 
+  private[zio] final def unsafeDone(r: ExitResult[E, A]): Unit = {
+    var retry: Boolean                = true
+    var joiners: List[Callback[E, A]] = null
+
+    while (retry) {
+      val oldState = state.get
+
+      val newState = oldState match {
+        case Pending(js) =>
+          joiners = js
+          Done(r)
+        case _ => oldState
+      }
+
+      retry = state.compareAndSet(oldState, newState)
```

Comment:
negate the boolean returned by the CAS ?

Reply:
woops 🤐